tasks/update_go.py

```diff
@@ -0,0 +1,171 @@
+) -> Dict[re.Pattern, str]:
+    patterns: Dict[re.Pattern, str] = {
+        re.compile(
+            "^(ARG GO_VERSION=)[.0-9]+$", flags=re.MULTILINE
+        ): rf"\g<1>{version}",
+    }
+    for (os, arch), sha in shas.items():
+        varname = f"GO_SHA256_{os.upper()}_{arch.upper()}"
+        pattern = re.compile(f"^(ARG {varname}=)[a-z0-9]+$", flags=re.MULTILINE)
+        replace = rf"\g<1>{sha}"
+
+        patterns[pattern] = replace
+
+    return patterns
+
+
+# returns a map from a pattern to what it should be replaced with, for windows file
+def _get_windows_patterns(
+    version: str, shas: Dict[Platform, str]
+) -> Dict[re.Pattern, str]:
+    patterns: Dict[re.Pattern, str] = {
+        re.compile(
+            '^(    "GO_VERSION"=")[.0-9]+(";)$', flags=re.MULTILINE
+        ): rf"\g<1>{version}\g<2>",
+    }
+    for (os, arch), sha in shas.items():
+        varname = f"GO_SHA256_{os.upper()}_{arch.upper()}"
+        pattern = re.compile(f'^(    "{varname}"=")[a-z0-9]+(";)$', flags=re.MULTILINE)
+        replace = rf"\g<1>{sha}\g<2>"
+
+        patterns[pattern] = replace
+
+    return patterns
+
+
+# returns the extension of the archive for the given os
+def _get_archive_extension(os: str) -> str:
+    if os == "windows":
+        return "zip"
+    return "tar.gz"
+
+
+# returns a map from platform to sha of the archive
+def _get_expected_sha256(version: str) -> Dict[Platform, str]:
+    # weirdly, the stored sha256 for round versions don't have a ".0" in the version
+    # while the archives have the ".0" suffix
+    version = version.removesuffix(".0")
+
+    shas: Dict[Platform, str] = {}
+    for os, arch in PLATFORMS:
+        ext = _get_archive_extension(os)
+        url = f"https://storage.googleapis.com/golang/go{version}.{os}-{arch}.{ext}.sha256"
+        res = requests.get(url)
+        sha = res.text.strip()
+        if len(sha) != 64:
+            raise exceptions.Exit(
+                f"The SHA256 of Go on {os}/{arch} has an unexpected format: '{sha}'"
+            )
+        shas[(os, arch)] = sha
+    return shas
+
+
+# checks that the archive sha is the same as the given one
+def _check_archive(version: str, shas: Dict[Platform, str]):
+    for (os, arch), expected_sha in shas.items():
+        ext = _get_archive_extension(os)
+        url = f"https://go.dev/dl/go{version}.{os}-{arch}.{ext}"
+        print(f"[check-archive] Fetching archive at {url}", file=sys.stderr)
+        # Using `curl` through `ctx.run` takes way too much time due to the archive being huge
+        # use `requests` as a workaround
+        req = requests.get(url)
+        sha = hashlib.sha256(req.content).hexdigest()
+        if sha != expected_sha:
+            raise exceptions.Exit(
+                f"The SHA256 of Go on {os}/{arch} should be {expected_sha}, but got {sha}"
+            )
+
+
+# replace patterns in a file
+def _handle_file(path: str, patterns: Dict[re.Pattern, str], expected_match: int = 1):
```

Comment:
hopefully we will have a possibility to share this within a module between repositories soon